File: Exercise_2.py
# Problem2 (https://leetcode.com/problems/reorder-list/)
# // Time Complexity : O(n)
# // Space Complexity :O(1)
# // Did this code successfully run on Leetcode :Yes
# // Any problem you faced while coding this :No


# // Your code here along with comments explaining your approach

def reverseSecondHalf(head):
    prev = None
    curr = head.next
    fast = curr.next
    while fast:
        curr.next = prev
        prev = curr
        curr = fast
        fast = fast.next
    curr.next = prev
    return curr


def reorderList(head):
    if head is None or head.next is None or head.next.next is None:
        return head

    # finding middle element
    slow = head
    fast = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
    # slow is middle element

    # reversing second half
    prev = None
    curr = slow.next
    # splitting the second half
    slow.next = None
    while curr:
        fast = curr.next
        curr.next = prev
        prev = curr
        curr = fast

    # merge 2 halves
    ptr1 = head
    ptr2 = prev
    while ptr2:
        temp = ptr1.next
        ptr1.next = ptr2
        ptr2 = ptr2.next
        ptr1.next.next = temp
        temp = temp.next

File: Exercise_2.cpp
// Time Complexity                              :  O(N/2) = O(N), where N is the number of nodes in the list
// Space Complexity                             :  O(1) - if we follow the iterative approach and 
//                                                 O(N/2) if we reverse the list recursively. 
// Did this code successfully run on Leetcode   :  Yes
// Any problem you faced while coding this      :  No

#include <bits/stdc++.h> 
using namespace std;  


// https://leetcode.com/problems/reorder-list/submissions/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
private: 
    ListNode *rhead;
public:
    void reorderList(ListNode* head) {
     
        ListNode *fast = head, *slow = head, *start = head;
        rhead = NULL;
        
        // O(N/2) = O(N)
        while(fast != NULL and fast->next != NULL) {
            fast = fast->next->next;
            slow = slow->next;
        }
        
        // O(N/2) operation
        // fast = reverse(slow) 
        reverseIterative(slow);
        
        // reorder the list - O(N/2)
        while(rhead->next != NULL) {
            ListNode *temp = start->next;
            start->next = rhead;
            start = temp;
            
            temp = rhead->next;
            rhead->next = start;
            rhead = temp;    
        }
    }
    
    // recursively reverse the list - O(N/2)
    ListNode* reverse(ListNode *head) {
        if(head->next == NULL) {
            rhead = head;
            return head;
        }
        
        ListNode *node = reverse(head->next);
        node->next = head;
        head->next = NULL;
        return head;
    }
    
    // iteratively reverse the list - O(N/2)
    void reverseIterative(ListNode *head) {
        if(!head) return;
        ListNode *prev = NULL, *curr = head;
        while(curr != NULL) {
            if(curr->next == NULL) rhead = curr;
            ListNode *temp = curr->next;
            curr->next = prev;
            prev = curr;
            curr = temp;
        }
    }
};